tests/phpunit/includes/testcase-canonical.php

```diff
@@ -1,54 +1,86 @@
 <?php
 
-class PLL_Canonical_UnitTestCase extends PLL_UnitTestCase {
-
-	// mainly copy paste from WP_Canonical_UnitTestCase::assertCanonical
-	public function assertCanonical( $test_url, $expected ) {
-		if ( is_string( $expected ) ) {
-			$expected = array( 'url' => $expected );
-		} elseif ( is_array( $expected ) && ! isset( $expected['url'] ) && ! isset( $expected['qv'] ) ) {
-			$expected = array( 'qv' => $expected );
-		}
+class PLL_Canonical_UnitTestCase extends WP_Canonical_UnitTestCase {
+	use PLL_UnitTestCase_Trait;
```

Comment:
You can't do this., as this is causing this error:
`PHP Warning:  Declaration of PLL_UnitTestCase_Trait::wpSetUpBeforeClass() should be compatible with WP_Canonical_UnitTestCase::wpSetUpBeforeClass($factory) in /home/travis/build/polylang/polylang/tests/phpunit/includes/testcase-canonical.php on line 17`

Reply:
I think we should adapt our `PLL_UnitTestCase_Trait`, as all of our test cases will receive the `$factory` parameter: https://github.com/WordPress/wordpress-develop/blob/86fc3af215a61f042804e3feb57627976d2bb54d/tests/phpunit/includes/abstract-testcase.php#L76

Reply:
I've faced some technical difficulties here:

-  We can't statically call `PLL_UnitTestCase_Trait::wpSetUpBeforeClass()` directly, as the `self` keyword would not resolve to `PLL_Canonical_UnitTestCase`, so all the needed properties will be set to `null`.
-  It is not possible to use [PHP traits' conflict resolution mechanism](https://www.php.net/manual/en/language.oop5.traits.php#language.oop5.traits.conflict) to override a method that doesn't have the same parameters.
- We can't use a `WP_Canonical_UnitTestCase` instance to delegate calls to `assertCanonical( $test_url, $expected_url )` as it would not let us override the `WP_Canonical_UnitTestCase::getCanonical( $test_url )` as [it is done here](https://github.com/polylang/polylang/blob/2565189a88c65ff30cb70890d50821148526be1b/tests/phpunit/includes/testcase-canonical.php#L77-L85)

This leaves us with the following solutions:

1. Adapting `PLL_UnitTestCase_Trait::wpSetUpBeforeClass()` to be compatible with `WP_Canonical_UnitTestCase::wpSetUpBeforeClass( $factory )` would force us to also adapt every classes that uses this trait, in `polylang`, `polylang-pro` and `polylang-woocommerce` as well.
2. Declaring a `PLL_Canonical_UnitTestCase::wpSetUpBeforeClass( $factory )` method to override `PLL_UnitTestCase::wpSetUpBeforeClass()` according to [PHP traits' precedence rules](https://www.php.net/manual/en/language.oop5.traits.php#language.oop5.traits.precedence) would force us to duplicate the code of `PLL_UnitTestCase_Trait::wpSetUpBeforeClass()` inside the new method.
3. Do not extend the `WP_Canonical_UnitTestCase` class and duplicate its `assertCanonical( $test_url, $expected_url )` method [as it was done before this PR](https://github.com/polylang/polylang/blob/0c9fd403caa735875c2a5dad077948f8a03f1103/tests/phpunit/includes/testcase-canonical.php#L6-L53).

i would rather not to resort to duplication whenever possible.